File: lib/src/internal/serializer/first_layer_style_settings_serializer.dart
import 'package:usercentrics_sdk/src/internal/internal.dart';
import 'package:usercentrics_sdk/src/model/first_layer_style_settings.dart';

class FirstLayerStyleSettingsSerializer {
  static dynamic serialize(FirstLayerStyleSettings? value) => value == null
      ? null
      : {
          'layout': LayoutSerializer.serialize(value.layout),
          'headerImage': _serializeHeader(value.headerImage),
          'title': _serializeTitle(value.title),
          'message': _serializeMessage(value.message),
          'buttonLayout': value.buttonLayout?.buttons
              .map(
                (row) => row
                    .map((button) => ButtonSettingsSerializer.serialize(button))
                    .toList(),
              )
              .toList(),
          'backgroundColor': ColorSerializer.serialize(value.backgroundColor),
          'overlayColor': ColorSerializer.serialize(value.overlayColor),
          'cornerRadius': value.cornerRadius,
        };

  static dynamic _serializeHeader(HeaderImageSettings? value) => value == null
      ? null
      : {
          'image': value.imageAssetPath,
          'height': value.height,
          'isExtended': value.isExtended,
          'isHidden': value.isHidden,
          'alignment': _serializeAlignment(value.alignment),
        };

  static dynamic _serializeTitle(TitleSettings? value) => value == null
      ? null
      : {
          'alignment': _serializeAlignment(value.alignment),
          'fontAssetPath': value.fontAssetPath,
          'textColor': ColorSerializer.serialize(value.textColor),
          'textSize': value.textSize,
        };

  static dynamic _serializeMessage(MessageSettings? value) {
    return value == null
        ? null
        : {
            'alignment': _serializeAlignment(value.alignment),
            'fontAssetPath': value.fontAssetPath,
            'textColor': ColorSerializer.serialize(value.textColor),
            'linkTextColor': ColorSerializer.serialize(value.linkTextColor),
            'linkTextUnderline': value.linkTextUnderline,
            'textSize': value.textSize,
          };
  }

  static dynamic _serializeAlignment(SectionAlignment? value) {
    switch (value) {
      case null:
        return null;
      case SectionAlignment.start:
        return 'START';
      case SectionAlignment.center:
        return 'CENTER';
      case SectionAlignment.end:
        return 'END';
    }
  }
}

File: android/src/main/kotlin/com/usercentrics/sdk/flutter/serializer/FirstLayerStyleSettingsSerializer.kt
package com.usercentrics.sdk.flutter.serializer

import com.usercentrics.sdk.*
import com.usercentrics.sdk.flutter.api.FlutterActivityProvider
import com.usercentrics.sdk.flutter.api.FlutterAssetsProvider

internal fun Any?.deserializeFirstLayerStyleSettings(
    assetsProvider: FlutterAssetsProvider,
    activityProvider: FlutterActivityProvider,
): FirstLayerStyleSettings? {
    if (this !is Map<*, *>) return null
    val layout = this["layout"].deserializeLayout()

    val headerImage =
        this["headerImage"]?.deserializeHeaderImageSettings(assetsProvider, activityProvider)

    val title = this["title"]?.deserializeTitleSettings(assetsProvider, activityProvider)
    val message = this["message"]?.deserializeMessageSettings(assetsProvider, activityProvider)

    val buttons = (this["buttonLayout"] as? List<*>)?.mapNotNull { row ->
        (row as? List<*>)?.mapNotNull {
            it.deserializeButtonSettings(
                assetsProvider,
                activityProvider
            )
        }
    }
    val buttonLayout = if (buttons != null) ButtonLayout.Grid(buttons) else null

    val backgroundColor = (this["backgroundColor"] as? String)?.deserializeColor()
    val overlayColor = (this["overlayColor"] as? String)?.deserializeColor()
    val cornerRadius = this["cornerRadius"] as? Int

    return FirstLayerStyleSettings(
        layout = layout,
        headerImage = headerImage,
        title = title,
        message = message,
        buttonLayout = buttonLayout,
        backgroundColor = backgroundColor,
        overlayColor = overlayColor,
        cornerRadius = cornerRadius,
    )
}

internal fun Any?.deserializeHeaderImageSettings(
    assetsProvider: FlutterAssetsProvider,
    activityProvider: FlutterActivityProvider,
): HeaderImageSettings? {
    if (this !is Map<*, *>) return null

    val isHidden = this["isHidden"] as? Boolean ?: false
    if (isHidden) {
        return HeaderImageSettings.Hidden
    }

    val image = this["image"].deserializeImage(assetsProvider, activityProvider) ?: return null

    val isExtended = this["isExtended"] as? Boolean ?: false
    return if (isExtended) {
        HeaderImageSettings.ExtendedLogoSettings(image)
    } else {
        HeaderImageSettings.LogoSettings(
            image = image,
            alignment = this["alignment"]?.deserializeSectionAlignment(),
            heightInDp = (this["height"] as? Number)?.toFloat()
        )
    }
}

internal fun Any?.deserializeSectionAlignment(): SectionAlignment? {
    if (this !is String) return null
    return SectionAlignment.valueOf(this)
}

internal fun Any?.deserializeTitleSettings(
    assetsProvider: FlutterAssetsProvider,
    activityProvider: FlutterActivityProvider,
): TitleSettings? {
    if (this !is Map<*, *>) return null

    return TitleSettings(
        alignment = this["alignment"]?.deserializeSectionAlignment(),
        font = this["fontAssetPath"].deserializeTypeface(assetsProvider, activityProvider),
        textColor = (this["textColor"] as? String)?.deserializeColor(),
        textSizeInSp = (this["textSize"] as? Number)?.toFloat(),
    )
}

internal fun Any?.deserializeMessageSettings(
    assetsProvider: FlutterAssetsProvider,
    activityProvider: FlutterActivityProvider,
): MessageSettings? {
    if (this !is Map<*, *>) return null

    return MessageSettings(
        alignment = this["alignment"]?.deserializeSectionAlignment(),
        font = this["fontAssetPath"].deserializeTypeface(assetsProvider, activityProvider),
        textColor = (this["textColor"] as? String)?.deserializeColor(),
        linkTextColor = (this["linkTextColor"] as? String)?.deserializeColor(),
        underlineLink = this["linkTextUnderline"] as? Boolean,
        textSizeInSp = (this["textSize"] as? Number)?.toFloat(),
    )
}
